File: backend/models/application.py
import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from db.base import Base, TimestampMixin


class Application(Base, TimestampMixin):
    __tablename__ = "applications"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("gen_random_uuid()"),
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    job_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("jobs.id", ondelete="SET NULL"),
    )
    # queued, pending_review, in_progress, applied, failed, skipped, withdrawn
    status: Mapped[str] = mapped_column(String, nullable=False, default="queued", index=True)
    applied_at: Mapped[datetime | None] = mapped_column()
    resume_used_url: Mapped[str | None] = mapped_column(String)
    cover_letter_used: Mapped[str | None] = mapped_column(Text)
    screenshot_url: Mapped[str | None] = mapped_column(String)
    error_message: Mapped[str | None] = mapped_column(Text)
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB)
    current_phase: Mapped[str | None] = mapped_column(String, default=None)
    phase_message: Mapped[str | None] = mapped_column(String, default=None)
    generated_application: Mapped[dict | None] = mapped_column("generated_application", JSONB, default=None)
    task_mode: Mapped[str | None] = mapped_column(String, default=None)

    user: Mapped["User"] = relationship(back_populates="applications")  # type: ignore[name-defined]  # noqa: F821
    job: Mapped["Job"] = relationship(lazy="noload")  # type: ignore[name-defined]  # noqa: F821

File: frontend/src-v2/types/application.ts
import type { Job } from "./job";

export interface GeneratedApplicationField {
  name: string;
  label: string;
  field_type: "text" | "select" | "checkbox" | "textarea" | "radio" | "file";
  options: { value: string; label: string }[];
  is_required: boolean;
  page_number: number;
}

export interface GeneratedApplicationAnswer {
  field_name: string;
  value: string;
  source: string;
}

export interface GeneratedApplication {
  task_id?: string | null;
  job_url?: string | null;
  ats_name?: string | null;
  pages_found: number;
  created_at?: string | null;
  fields: GeneratedApplicationField[];
  answers: GeneratedApplicationAnswer[];
}

export interface Application {
  id: string;
  user_id: string;
  job_id?: string | null;
  status: string;
  applied_at?: string | null;
  resume_used_url?: string | null;
  cover_letter_used?: string | null;
  screenshot_url?: string | null;
  error_message?: string | null;
  created_at: string;
  current_phase?: string | null;
  phase_message?: string | null;
  generated_application?: GeneratedApplication | null;
  task_mode?: string | null;
  job?: Job | null;
}

export interface ApplicationListResponse {
  applications: Application[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface ApplicationStats {
  total: number;
  applied: number;
  pending: number;
  failed: number;
  skipped: number;
  this_week: number;
  success_rate: number;
}
